tenantcostserver: basic token bucket implementation

RaduBerinde opened the pull request:
#### roachpb: add more fields to TokenBucket APIs

Release note: None

#### tenantcostserver: basic token bucket implementation

This commit adds a basic implementation of the server-side token
bucket. The limitations are that we don't use shares and we don't
distinguish between different instances. We also don't implement a
burst limit yet - the current plan from the CC side is that this will
never be set (even for free tier).

This logic doesn't do much until we add the corresponding client-side
limiting code, which will happen in a future change. At that point,
this server-side code should be good enough for a single instance (and
not too bad even for multiple instances: the debt mechanism will limit
overuse in the long term).

Release justification: category 2, Serverless only code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69274)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, with a few questions/nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-:-MhsYkrf3GzstdgOQPP2:bxs0i52)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsV1AX7kM8JKGCrJp9:-MhsV1AX7kM8JKGCrJpA:b9dikkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a25bab893c68d8e94806f3b2cd828ee9a19c6bd9/pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go#L81)):*
> ```Go
> 			debtRate := debt / req.TargetRequestPeriod.Seconds()
> 			availableRate -= debtRate
> 			availableRate = math.Max(availableRate, 0.01*s.RURefillRate)
> ```

Shouldn't this be `.1` if you want to use at most 90% of the rate for the debt? If not, add a bit to the comment explaining why.
___
*[pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/testdata/request, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsVQSd3NYQ0sFl5CBN:-MhsVQSd3NYQ0sFl5CBO:b-50maqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a25bab893c68d8e94806f3b2cd828ee9a19c6bd9/pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/testdata/request#L28)):*
> ```
> Granted RUs: 200
> Trickle duration: 1s
> Current RUs: -200
> ```

If we had `100` left, and then requested `200`, shouldn't it be `-100`? If `-200` is correct, add a comment explaining why, since it's surprising.
___
*[pkg/roachpb/api.proto, line 2512 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsSJCU4Yi7hQyqGGWs:-MhsSJCV9Ooz3mO3Vehu:bpd412d) ([raw file](https://github.com/cockroachdb/cockroach/blob/14a008702f947b1c533006edb5fcf748a2ff3936/pkg/roachpb/api.proto#L2512)):*
> ```protobuf
>   // trickle time period. E.g. if we are granted 1000 tokens and the trickle
>   // duration is 10 seconds, the tokens become available at a rate of
>   // 100tokens/sec for the next 10 seconds.
> ```

NIT: 100 tokens/sec
___
*[pkg/roachpb/api.proto, line 2513 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsSPJX6MC4R7HP8I6X:-MhsSPJX6MC4R7HP8I6Y:biepa2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/14a008702f947b1c533006edb5fcf748a2ff3936/pkg/roachpb/api.proto#L2513)):*
> ```protobuf
>   // duration is 10 seconds, the tokens become available at a rate of
>   // 100tokens/sec for the next 10 seconds.
>   // TrickleTime is at most the given TargetRequestPeriod.
> ```

NIT: TrickleTime => TrickleDuration


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-:-MhtxvP17uLIHDs77ixL:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @andy-kimball)

___
*[pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsV1AX7kM8JKGCrJp9:-MhtvzeFByoWm1-8G2fC:b4t0olj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a25bab893c68d8e94806f3b2cd828ee9a19c6bd9/pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go#L81)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Shouldn't this be `.1` if you want to use at most 90% of the rate for the debt? If not, add a bit to the comment explaining why.
</blockquote></details>

Ha, I think I dithered between 90% and 99%. Made it 95% :)
___
*[pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/testdata/request, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69274#-MhsVQSd3NYQ0sFl5CBN:-MhtwA0p4JuZtF-PZoMs:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a25bab893c68d8e94806f3b2cd828ee9a19c6bd9/pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/testdata/request#L28)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If we had `100` left, and then requested `200`, shouldn't it be `-100`? If `-200` is correct, add a comment explaining why, since it's surprising.
</blockquote></details>

Good catch, fixed.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
tenantcostserver: basic token bucket implementation

This commit adds a basic implementation of the server-side token
bucket. The limitations are that we don't use shares and we don't
distinguish between different instances. We also don't yet implement a
burst limit yet - the current plan from the CC side is that this will
never be set (even for free tier).

This logic doesn't do much until we add the corresponding client-side
limiting code, which will happen in a future change. At that point,
this server-side code should be good enough for a single instance (and
not too bad even for multiple instances: the debt mechanism will limit
overuse in the long term).

Release justification: category 2, Serverless only code.

Release note: None

RaduBerinde committed:
roachpb: add more fields to TokenBucket APIs

Release note: None
